Validation refactoring

voidxnull opened the pull request:
Based on the `uikit-port` branch. Contains a lot of fixes to the new webpack config. Once merged, it'll probably break the alpha (I don't have an access to the server so I can't say for sure).

- [x] Replace checkit validations with joi
- [x] Refactor post validations.
  - [x] Test new post validations.
- [x] Refactor ProfilePost to not validate on save
- [x] Fix tests

voidxnull committed:
Replace checkit with Joi for server-side validations

indeyets committed:
Move react-middleware into separate module

indeyets committed:
Upgrade NodeJS

indeyets committed:
Fix recursive definition of Geotag proptype

indeyets committed:
Compile server using webpack

indeyets committed:
Add initial UIKit implementation

indeyets committed:
Fix tests

indeyets committed:
Lock dependencies

artkravchenko committed:
Get the test specs to work

voidxnull committed:
Update yarn.lock

voidxnull committed:
Update yarn.lock

indeyets committed:
Add support for router's "change" callbacks

voidxnull committed:
Fix path to webpack-chunks.json in getReactMiddleware for tests

voidxnull committed:
Refactor Post validation

voidxnull committed:
Temporarily fix tests

Disabled some tests that didn't work without compiling.

voidxnull committed:
Update node to 8.4

voidxnull committed:
Fix webpack configs

voidxnull committed:
Fix travis

voidxnull committed:
Fix eslint errors

artkravchenko reviewed (changes requested):
The PR doesn't contain a production server start command, like `npm run docker-start` in the past.

However, the PR hasn't been tested enough with production mode enabled. After some research, I found blocking issues with `compression-webpack-plugin` and `zopfli`, server js oriented `babel-loader`'s list of plugins, paths to js files potentially attached to `ejs` template, `babel-plugin-react-remove-prop-types` and few others.

The fixes are almost ready. Today I'm going to finish the debugging and push the appropriate commits.

In addition, the development mode is not optimized and it has low performance and high RAM usage. Unfortunately, I don't know the exact reasons about it. It seems we maybe should break down bunch of new code and go with smaller steps to catch the problem.

artkravchenko committed:
Create a short command to start server with production mode enabled

artkravchenko committed:
Get rid of plugins array's falsy entries in the server js babel-loader options

artkravchenko committed:
Include ExtractTextPlugin only in the client build process

artkravchenko committed:
Use zopfli-webpack-plugin to compress the assets

compression-webpack-plugin@1.0.0 comes with breaking change removing node-zopfli integration

Add a note to the repository's README about Zopfli dependency

artkravchenko committed:
Attach correct paths of js assets to the ejs template

artkravchenko committed:
Disable babel-plugin-react-remove-prop-types to prevent the errors

'TypeError: Cannot convert undefined or null to object' is thrown if the prop types of certain components are removed

artkravchenko committed:
Edit the note about version of NodeJS used by the development team
